### ISSUE TITLE
Cannot find symbol _start for building optee_os

### ISSUE THREAD

**Issue by gitgonewithwind** [OP-TEE/optee_os#3418]
I am following the guide here https://optee.readthedocs.io/en/latest/building/gits/optee_os.html#optee-os to build optee_os
but error ouput:

make     CFG_TEE_BENCHMARK=n     CFG_TEE_CORE_LOG_LEVEL=3     CROSS_COMPILE=arm-linux-gnueabi-     CROSS_COMPILE_core=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm32=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm64=aarch64-linux-gnu-     DEBUG=1     O=out/arm     PLATFORM=vexpress-qemu_virt
  CHK     out/arm/conf.mk
  CHK     out/arm/include/generated/conf.h
  GEN     out/arm/core/tee-header_v2.bin
Cannot find symbol _start
core/arch/arm/kernel/link.mk:191: recipe for target 'out/arm/core/tee-header_v2.bin' failed
make: *** [out/arm/core/tee-header_v2.bin] Error 1
make: *** Deleting file 'out/arm/core/tee-header_v2.bin'

Note: I changed gcc from arm-linux-gnueabihf- to arm-linux-gnueabi-  (I follow the toolchain guide to download, the name is arm-linux-gnueabi-)

**Comment by vchong** [comment]
I've the same issue right now on hikey aosp builds.

```
Cannot find symbol __data_end
Cannot find symbol _start
Cannot find symbol __init_size
```

gcc on aosp builds is different too.

**Comment by vchong** [comment]
@gitgonewithwind Can you run `readelf -a tee.elf` and paste the lines that contain `__init_size`, `__data_end` and `_start` please? They'll look something like below:

```
4193: 000000000006a6c0     0 NOTYPE  GLOBAL DEFAULT  ABS __init_size
3879: 000000003f06a6c0     0 NOTYPE  LOCAL  DEFAULT    5 __data_end
4423: 000000003f000000   256 FUNC    GLOBAL DEFAULT    1 _start
```

Thanks

**Comment by gitgonewithwind** [comment]
after running readelf -a tee.elf , I got a long list. something is likes below:(but now  I do not  compile it individually , I am using yocto to build optee successfully, so maybe I give up this issue here,do not debug it)  
...
0x9e000000 <_start>: 0x1 [cantunwind]


0057: 00085350     0 NOTYPE  GLOBAL DEFAULT  ABS __init_size
...
 10330: 9e085350     0 NOTYPE  GLOBAL DEFAULT    6 __data_end

there is NO _start

**Comment by gitgonewithwind** [comment]
the updated version in yocto has the same issue. need fix it

**Comment by jenswi-linaro** [comment]
Should be fixed by https://github.com/OP-TEE/optee_os/pull/3425

**Comment by jforissier** [comment]
> Should be fixed by #3425

Yep, and that's commit f77987ae8c99 in the master branch. @gitgonewithwind please check if you have this commit in your repo.

**Comment by gitgonewithwind** [comment]
> > Should be fixed by #3425
> 
> Yep, and that's commit [f77987a](https://github.com/OP-TEE/optee_os/commit/f77987ae8c99f93a42f2085cf44c4a3116386b57) in the master branch. @gitgonewithwind please check if you have this commit in your repo.

==================
thanks, I saw there was merged 6 hours ago, but just now, in yocto I compiled for the updated master version. the issue  is still  there , let's me try again
??

**Comment by gitgonewithwind** [comment]
sorry, I tried again in yocto for master updated version or commit f77987a(merged 6 hours ago,I found it in git log) both just now. same issue. let me try again.

**Comment by jforissier** [comment]
@gitgonewithwind which compiler are you using exactly? Please post the output of `arm-linux-gnueabi-gcc --version`.

**Comment by gitgonewithwind** [comment]
> @gitgonewithwind which compiler are you using exactly? Please post the output of `arm-linux-gnueabi-gcc --version`.


I checked my yocto, gcc is gcc-7.3.0

**Comment by jforissier** [comment]
> is it too old?

As the French saying goes: "Best soups are made in old pots". :smile: 

I've tried 3 different compilers and could not reproduce the issue. Please post detailed instructions, if possible without using yocto -- in your first message you say you reproduce the issue when trying to build inside `optee_os` for `PLATFORM=vexpress-qemu_virt`? How so?

**Comment by gitgonewithwind** [comment]
now I do not use yocto, and use updated optee-os, ubuntu 16.04 in virtalbox, 
toolchains download at here guide a few days ago: https://optee.readthedocs.io/en/latest/building/toolchains.html#toolchains
different issue appear: 

Thank you, let me take my time to fix it. 

=========
arm-linux-gnueabi-gcc --versionarm-linux-gnueabi-gcc (GNU Toolchain for the A-profile Architecture 8.2-2019.01 (arm-rel-8.28)) 8.2.1 20180802
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
---
make FG_TEE_BENCHMARK=n     CFG_TEE_CORE_LOG_LEVEL=3     CROSS_COMPILE=arm-linux-gnueabi-     CROSS_COMPILE_core=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm32=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm64=aarch64-linux-gnu-     DEBUG=1     O=out/arm     PLATFORM=vexpress-qemu_virt
  CHK     out/arm/conf.mk
  CHK     out/arm/include/generated/conf.h
  CHK     out/arm/core/include/generated/asm-defines.h
  GEN     out/arm/core/ldelf_hex.c

***
Can't find elftools module. Probably it is not installed on your system.
You can install this module with

$ apt install python3-pyelftools

if you are using Ubuntu. Or try to search for "pyelftools" or "elftools" in
your package manager if you are using some other distribution.
***

Traceback (most recent call last):
  File "scripts/gen_ldelf_hex.py", line 13, in <module>
    from elftools.elf.elffile import ELFFile
ModuleNotFoundError: No module named 'elftools'
mk/subdir.mk:160: recipe for target 'out/arm/core/ldelf_hex.c' failed
make: *** [out/arm/core/ldelf_hex.c] Error 1
-----


I checkouted
sudo apt install python3-pyelftools
Reading package lists... Done
Building dependency tree       
Reading state information... Done
python3-pyelftools is already the newest version (0.23-2).
0 upgraded, 0 newly installed, 0 to remove and 349 not upgraded

**Comment by gitgonewithwind** [comment]
I change to ubuntu 16.04 true hardware machine (not virtualbox)
toolchains download at same guide.
optee_os commit
commit 33017d856e8c30b4a0c550d636842409a0a3f6c2
Author: Jerome Forissier <jerome@forissier.org>
Date:   Fri Nov 22 15:34:20 2019 +0100

testing result:

~/work/optee_os$ arm-linux-gnueabi-gcc --version
arm-linux-gnueabi-gcc (GNU Toolchain for the A-profile Architecture 8.2-2019.01 (arm-rel-8.28)) 8.2.1 20180802
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

compling optee-os

make FG_TEE_BENCHMARK=n     CFG_TEE_CORE_LOG_LEVEL=3     CROSS_COMPILE=arm-linux-gnueabi-     CROSS_COMPILE_core=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm32=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm64=aarch64-linux-gnu-     DEBUG=1     O=out/arm     PLATFORM=vexpress-qemu_virt
  CHK     out/arm/conf.mk
  CHK     out/arm/include/generated/conf.h
  GEN     out/arm/core/tee-header_v2.bin
Cannot find symbol _start
core/arch/arm/kernel/link.mk:191: recipe for target 'out/arm/core/tee-header_v2.bin' failed
make: *** [out/arm/core/tee-header_v2.bin] Error 1
make: *** Deleting file 'out/arm/core/tee-header_v2.bin'

**Comment by gitgonewithwind** [comment]
I think the toolchians in my just upper individual test is not old, the issue is still there. 
anyway maybe my yocto toolchains is old. 
but why in my testing in just upper, the issue is same as in my yocto compiling

**Comment by jforissier** [comment]
I just tried the same compiler as yours, same commit, same build command line: no error here (Ubuntu 19.10).
Can you please paste the output of `readelf -s out/arm/core/tee.elf | grep -w _start`? I do have this symbol in my `tee.elf`:
```
$ readelf -s out/arm/core/tee.elf | grep -w _start
  9776: 0e100000   148 FUNC    GLOBAL DEFAULT    1 _start
```

**Comment by gitgonewithwind** [comment]
readelf -s out/arm/core/tee.elf | grep -w _start
 10315: 9e000000   144 FUNC    GLOBAL DEFAULT    1 _start

**Comment by gitgonewithwind** [comment]
I tried again,same issue.

**Comment by jforissier** [comment]
@gitgonewithwind looks good, as far as I can tell. I suspect the problem is with the `elftools` Python module in Ubuntu 16.04, or the way we use it in `scripts/gen_tee_bin.py`. Perhaps you could add debug traces in the `get_symbol()` function?

**Comment by gitgonewithwind** [comment]
> @gitgonewithwind looks good, as far as I can tell. I suspect the problem is with the `elftools` Python module in Ubuntu 16.04, or the way we use it in `scripts/gen_tee_bin.py`. Perhaps you could add debug traces in the `get_symbol()` function?

--
I am not familiar  py debug traces in the get_symbol() , let me check online, or please tell how to add something before get_symbol() in scripts/gen_tee_bin.py

**Comment by eozturk1** [comment]
I have the same problem with symbols on Ubuntu 16.04 with rpi3.xml. optee_os has the commit f77987a. Trace:

```
Cannot find symbol __init_size
core/arch/arm/kernel/link.mk:197: recipe for target 'out/arm/core/tee-pager_v2.bin' failed
make[1]: *** [out/arm/core/tee-pager_v2.bin] Error 1
make[1]: *** Deleting file 'out/arm/core/tee-pager_v2.bin'
make[1]: *** Waiting for unfinished jobs....
Cannot find symbol __init_size
core/arch/arm/kernel/link.mk:204: recipe for target 'out/arm/core/tee-pageable_v2.bin' failed
make[1]: *** [out/arm/core/tee-pageable_v2.bin] Error 1
make[1]: *** Deleting file 'out/arm/core/tee-pageable_v2.bin'
Cannot find symbol _start
core/arch/arm/kernel/link.mk:191: recipe for target 'out/arm/core/tee-header_v2.bin' failed
make[1]: *** [out/arm/core/tee-header_v2.bin] Error 1
make[1]: *** Deleting file 'out/arm/core/tee-header_v2.bin'
make[1]: Leaving directory '/home/ercan/optee-project/optee_os'
common.mk:380: recipe for target 'optee-os-common' failed
make: *** [optee-os-common] Error 2
make: *** Waiting for unfinished jobs....
```

**Comment by jforissier** [comment]
@eozturk1 thanks for the information. I'll install 16.04 in a VM and see what's wrong with the Python script, unless someone beats me to it.

**Comment by eozturk1** [comment]
Thanks @jforissier ! Please let me know if you need any more information from me about the problem!

**Comment by jforissier** [comment]
https://github.com/OP-TEE/optee_os/pull/3430

**Comment by jforissier** [comment]
@gitgonewithwind @eozturk1 I'm closing this issue because I think it is resolved by the above PR which has been merged. Please re-open if you still have issues.

**Comment by gitgonewithwind** [comment]
@jforissier sorry bother you again:

I tried updated comm, another error coming again, before is OK. I wass confused. ubuntu 16.04  hardware host(not virtulbox)

commit d77929ec2243b789b58b718d7cd50da87cd9cf97
Author: Sumit Garg <sumit.garg@linaro.org>

Date:   Wed Nov 27 21:01:27 2019 +0530

    core: ftrace: dump core load address to support ASLR
    
    Additionally dump core address in ftrace buffer to support syscall
    tracing in case TEE core ASLR is enabled.
    
    Signed-off-by: Sumit Garg <sumit.garg@linaro.org>
    Reviewed-by: Jerome Forissier <jerome@forissier.org>
    [jf: s/Load address @/TEE load address @/]
    Signed-off-by: Jerome Forissier <jerome@forissier.org>


-----




make FG_TEE_BENCHMARK=n     CFG_TEE_CORE_LOG_LEVEL=3     CROSS_COMPILE=arm-linux-gnueabi-     CROSS_COMPILE_core=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm32=arm-linux-gnueabi-     CROSS_COMPILE_ta_arm64=aarch64-linux-gnu-     DEBUG=1     O=out/arm     PLATFORM=imx-mx7dpico_mbl
  CHK     out/arm/conf.mk
  CHK     out/arm/include/generated/conf.h
  GEN     out/arm/core/ta_pub_key.c
Traceback (most recent call last):
  File "scripts/pem_to_pub_c.py", line 61, in <module>
    main()
  File "scripts/pem_to_pub_c.py", line 24, in main
    from Crypto.PublicKey import RSA
ModuleNotFoundError: No module named 'Crypto'
mk/subdir.mk:160: recipe for target 'out/arm/core/ta_pub_key.c' failed
make: *** [out/arm/core/ta_pub_key.c] Error 1

**Comment by jforissier** [comment]
`sudo apt install python-pycrypto` perhaps?

**Comment by gitgonewithwind** [comment]
@jforissier  
thanks . let's me try. maybe I tried before.

**Comment by gitgonewithwind** [comment]
@jforissier 
Thanks. I fix this problem. (I do not know I installed what package to fix it ). 
now same problem in yocto project. I open another issue. I am compiling it to try 
https://github.com/OP-TEE/optee_os/issues/3441
